src/main/java/org/opensearch/neuralsearch/query/SparseEncodingQueryBuilder.java

```diff
@@ -0,0 +1,266 @@
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.builder.EqualsBuilder;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.apache.lucene.document.FeatureField;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.opensearch.common.SetOnce;
+import org.opensearch.core.ParseField;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.ParsingException;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryRewriteContext;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.neuralsearch.ml.MLCommonsClientAccessor;
+import org.opensearch.neuralsearch.util.TokenWeightUtil;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * SparseEncodingQueryBuilder is responsible for handling "sparse_encoding" query types. It uses an ML SPARSE_ENCODING model
+ * or SPARSE_TOKENIZE model to produce a Map with String keys and Float values for input text. Then it will be transformed
+ * to Lucene FeatureQuery wrapped by Lucene BooleanQuery.
+ */
+
+@Log4j2
+@Getter
+@Setter
+@Accessors(chain = true, fluent = true)
+@NoArgsConstructor
+@AllArgsConstructor
+public class SparseEncodingQueryBuilder extends AbstractQueryBuilder<SparseEncodingQueryBuilder> {
+    public static final String NAME = "sparse_encoding";
+    @VisibleForTesting
+    static final ParseField QUERY_TEXT_FIELD = new ParseField("query_text");
+    @VisibleForTesting
+    static final ParseField MODEL_ID_FIELD = new ParseField("model_id");
+
+    private static MLCommonsClientAccessor ML_CLIENT;
+
+    public static void initialize(MLCommonsClientAccessor mlClient) {
+        SparseEncodingQueryBuilder.ML_CLIENT = mlClient;
+    }
+
+    private String fieldName;
+    private String queryText;
+    private String modelId;
+    private Supplier<Map<String, Float>> queryTokensSupplier;
+
+    /**
+     * Constructor from stream input
+     *
+     * @param in StreamInput to initialize object from
+     * @throws IOException thrown if unable to read from input stream
+     */
+    public SparseEncodingQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.queryText = in.readOptionalString();
+        this.modelId = in.readOptionalString();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeOptionalString(queryText);
+        out.writeOptionalString(modelId);
```

Comment:
Why are these 2 optional? In the fromXContent it looks like they require a value.

Reply:
Good point. In original design there were 2 query forms and in final proposal we only keep this one. But I didn't change here. Will fix this.